--- ../../mp3gain/mpglibDBL/VbrTag.h	2003-08-12 02:02:55.000000000 +0200
+++ ../../mp3gain/mpglibDBL/VbrTag.h	2005-02-06 23:15:39.726386376 +0100
@@ -24,6 +24,13 @@
 #include "lame.h"
 #include "config.h"
 
+#if HAVE_INTTYPES_H
+ # include <inttypes.h>
+#else
+ # if HAVE_STDINT_H
+  #  include <stdint.h>
+ # endif
+#endif
 /* -----------------------------------------------------------
  * A Vbr header may be present in the ancillary
  * data field of the first frame of an mp3 bitstream
